### ISSUE TITLE
luz_callback_auto_resume doesn't resume ctx$epoch

### ISSUE THREAD

**Issue by pradosj** [mlverse/luz#136]
I'm trying to use luz_callback_auto_resume() to resume a model after a fail in the model training.
However, it seems that ctx$epoch is not resumed as I would imagine after a fail.
Indeed, if my training fail at epoch "n", after resuming, the epoch stored into ctx$epoch is 1.

Is it an expected behaviour ?
Can it be fixed ?

And thanks a lot, for this wonderful package.

Best,
J.

**Comment by dfalbel** [comment]
Hi @pradosj,

This is indeed not the expected behavior but they way it's implement might cause some confusion.
Basically, it will run trough all epochs from `1 to N` but will just do nothing in the training loops for the epochs it has already trained for, this handled by:

https://github.com/mlverse/luz/blob/f7e160f2140738a2c4ad63acb62fabc501c69027/R/callbacks-resume.R#L116-L121

However, depending on the callbacks orders, the `on_epoch_begin` for the autoresume callback might be running after another callback which can break your assumption. Maybe it works if you re-order the callbacks making sure that the au-resume callback is placed first in the callbacks list. 

I think we can also help with this by. setting a lower default `weight` value so it's by default executed before the other callbacks.

**Comment by pradosj** [comment]
Hi, @dfalbel 
Thank you for the answer.
So yes, ctx$epoch is resumed correctly, and I now realise my issue is coming because the epoch used by torch in lr_scheduler() is not ctx$epoch.

Basically, combining luz_callback_lr_scheduler() with luz_callback_auto_resume() is tricky. And the later doesn't necessarily resume the expected LR, as shown in example code below.  


`
library(torch)
library(luz)

# Callback to display learning rate at each epoch
luz_callback_lr_progress <- luz_callback(
  on_epoch_begin = function() {
    rlang::inform(glue::glue("lr={ctx$opt$param_groups[[1]]$lr}"))
  }
)

# Callback to simulate a learning issue
luz_callback_simulate_failure <- luz_callback(
  initialize = function(at_epoch) {
    self$at_epoch = at_epoch
  },
  on_epoch_begin = function() {
    if (ctx$epoch>=self$at_epoch) rlang::abort("simulated failure")
  }
)

# Train a linear model on iris updating the LR at each epoch and simulate a failure at epoch 5
# The LR start at 1 and is divided by 10 at each step so that LR is expected to be 1e-5 at epoch 5.
fm <- nn_linear |>
  setup(loss = nn_multi_margin_loss(),optimizer = optim_adam) |>
  set_opt_hparams(lr=1) |>
  set_hparams(in_features=4L,out_features=3L) |>
  fit(
    data = list(x=data.matrix(iris[1:4]),y=iris$Species),
    dataloader_options = list(shuffle=TRUE,batch_size=150L,drop_last=FALSE),
    callbacks = list(
      luz_callback_auto_resume("~/Downloads/state.pt"),
      luz_callback_lr_scheduler(lr_step,step_size=1L),
      luz_callback_simulate_failure(at_epoch=5L),
      luz_callback_lr_progress()
    )
  )

# Now resume model training after failure using the same code, but without failure.
# The issue after this resume is that LR computed for epoch 5 is not 1e-5 as expected but restart at 1 !
fm <- nn_linear |>
  setup(loss = nn_multi_margin_loss(),optimizer = optim_adam) |>
  set_opt_hparams(lr=1) |>
  set_hparams(in_features=4L,out_features=3L) |>
  fit(
    data = list(x=data.matrix(iris[1:4]),y=iris$Species),
    dataloader_options = list(shuffle=TRUE,batch_size=150L,drop_last=FALSE),
    callbacks = list(
      luz_callback_auto_resume("~/Downloads/state.pt"),
      luz_callback_lr_scheduler(lr_step,step_size=1L),
      luz_callback_simulate_failure(at_epoch=+Inf),
      luz_callback_lr_progress()
    )
  )
`

Any suggestion to prevent this behaviour ?

Many thanks 
Julien Prados

**Comment by dfalbel** [comment]
Hi, 

Thank you Julien, I can reproduce the problem and will make sure to fix it soon.

The first problem is that we didn't correctly handle state dicts from learning rate schedulers in torch (see https://github.com/mlverse/torch/pull/1100O) 

We will now make sure we handle them in luz too, via the `state_dict()` callbacks interface.

**Comment by dfalbel** [comment]
Hi @pradosj 

The problem should be fixed with #137. I you have time, please try it and let me know if it works for you.
You can install the dev version of luz with:

```
remotes::install_github("mlverse/luz")
```

It will also require the dev version of torch.
